{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# run hapBLOCK on diploid gt vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append('/mnt/archgen/users/yilei/tools/hapBLOCK/python3')\n",
    "from IO.h5_modify import merge_in_af, get_af, get_af1000G, lift_af, save_h5, get_af1000G_atGivenPos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "!for ch in {1..22};do bcftools view -r $ch -m2 -M2 -v snps -O v -o ./vcfbychr/chr$ch.temp.vcf merged.glimpse.GP99.MAC1.transversion_only.vcf.gz; done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import h5py\n",
    "import pandas as pd\n",
    "\n",
    "def merge_in_ld_map(path_h5, path_snp1240k, chs=range(1,23), write_mode=\"a\"):\n",
    "    \"\"\"Merge in MAP from eigenstrat .snp file into\n",
    "    hdf5 file. Save modified h5 in place \n",
    "    path_h5: Path to hdf5 file to modify.\n",
    "    path_snp1240k: Path to Eigenstrat .snp file whose map to use\n",
    "    chs: Which Chromosomes to merge in HDF5 [list].\n",
    "    write_mode: Which mode to use on hdf5. a: New field. r+: Change Field\"\"\"\n",
    "    with h5py.File(path_h5, \"r\") as f:\n",
    "        print(\"Lifting LD Map from eigenstrat to HDF5...\")\n",
    "        print(\"Loaded %i variants.\" % np.shape(f[\"calldata/GT\"])[0])\n",
    "        print(\"Loaded %i individuals.\" % np.shape(f[\"calldata/GT\"])[1])\n",
    "\n",
    "        ### Load Eigenstrat\n",
    "        df_snp = pd.read_csv(path_snp1240k, header=None, sep=r\"\\s+\", engine=\"python\")\n",
    "        df_snp.columns = [\"SNP\", \"chr\", \"map\", \"pos\", \"ref\", \"alt\"]\n",
    "\n",
    "        rec = np.zeros(len(f[\"variants/POS\"]))  # Create the array for vector\n",
    "\n",
    "        for ch in chs:\n",
    "            df_t = df_snp[df_snp[\"chr\"] == ch]\n",
    "            print(f\"Loaded {len(df_t)} Chr.{ch} 1240K SNPs.\")\n",
    "\n",
    "            idx_f = f[\"variants/CHROM\"][:].astype(\"str\")==str(ch)\n",
    "            if np.sum(idx_f)==0:  # If no markers found jump to next chromosome\n",
    "                print(\"Did not find any markers...\")\n",
    "                continue\n",
    "            rec_ch = np.zeros(len(idx_f), dtype=\"float\")\n",
    "\n",
    "            ### Intersect SNP positions\n",
    "            its, i1, i2 = np.intersect1d(f[\"variants/POS\"][idx_f], df_t[\"pos\"], return_indices=True)\n",
    "\n",
    "            l = np.sum(idx_f)\n",
    "            print(f\"Intersection {len(i2)} out of {l} HDF5 SNPs\")\n",
    "\n",
    "            ### Extract Map positions\n",
    "            rec_ch[i1] = df_t[\"map\"].values[i2]  # Fill in the values in Recombination map\n",
    "\n",
    "            ### Interpolate if Needed (map position still 0)\n",
    "            itp_idx = (rec_ch == 0)\n",
    "            if np.sum(itp_idx) > 0:   # In case we have to interpolate\n",
    "                print(f\"Interpolating {np.sum(itp_idx)} variants.\")\n",
    "                x = df_t[\"pos\"] \n",
    "                y = df_t[\"map\"]   \n",
    "                x1 = f[\"variants/POS\"][:][idx_f]  # Extract all positions of interest\n",
    "                rec_ch = np.interp(x1, x, y) \n",
    "            \n",
    "            ### Make sure that sorted\n",
    "            assert(np.all(np.diff(rec_ch)>=0))  # Assert the Recombination Map is sorted! (no 0 left and no funky stuff)\n",
    "            rec[idx_f]=rec_ch # Set the Map position for chromosome indices\n",
    "            print(f\"Finished Chromosome {ch}.\")\n",
    "    \n",
    "    ### Now create the new column in hdf5\n",
    "    print(\"Adding map to HDF5...\")\n",
    "    with h5py.File(path_h5, write_mode) as f0:\n",
    "        group = f0[\"variants\"]\n",
    "        l = len(f0[\"variants/POS\"])\n",
    "        if write_mode == \"a\":  # If appending new data\n",
    "            group.create_dataset('MAP', (l,), dtype='f')   \n",
    "        f0[\"variants/MAP\"][:] = rec[:]\n",
    "    print(\"We did it. Finished.\")\n",
    "\n",
    "import allel\n",
    "def vcf_to_1240K_hdf(in_vcf_path = \"/n/groups/reich/ali/WholeGenomeImputation/imputed/v43.4/chr3.bcf\",\n",
    "                     path_vcf = \"./data/vcf/1240k_v43/ch3.vcf.gz\",\n",
    "                     path_h5 = \"./data/hdf5/1240k_v43/ch3.h5\",\n",
    "                     map_path=\"/n/groups/reich/DAVID/V43/V43.5/v43.5.snp\",\n",
    "                     ch=3):\n",
    "    \"\"\"Convert Ali's vcf to 1240K hdf5. \n",
    "    If marker_path empty, no SNP filtering done.\n",
    "    If map_path empty, no genetic map is merged in.\n",
    "    \"\"\" \n",
    "    path_vcf = in_vcf_path # Use the unfiltered input in next step\n",
    "    \n",
    "    allel.vcf_to_hdf5(input=path_vcf, output=path_h5, \n",
    "                  fields = ['variants/*', 'calldata/*', \"samples\"], compression=\"gzip\") # Do the conversion to hdf5. Takes hours\n",
    "    print(\"Finished conversion to hdf5!\")\n",
    "    \n",
    "    if len(map_path)>0:\n",
    "        merge_in_ld_map(path_h5=path_h5, \n",
    "                    path_snp1240k=map_path,\n",
    "                    chs=[ch])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for ch in range(1,23):\n",
    "    \n",
    "    in_vcf_path = f\"./vcfbychr/chr{ch}.temp.vcf\"\n",
    "    path_h5 = f\"./hdf5/ch{ch}.h5\"\n",
    "    map_path = f\"/mnt/archgen/users/hringbauer/data/MinMyc.snp\"\n",
    "    \n",
    "    vcf_to_1240K_hdf(in_vcf_path = in_vcf_path, path_vcf = \"\", path_h5=path_h5,\n",
    "                    map_path = map_path, ch=ch)\n",
    "\n",
    "    # grabbing 1000G AF\n",
    "    # with h5py.File(f'/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/chr{ch}.hdf5', 'r') as f_ref:\n",
    "    #     with h5py.File(path_h5, 'r') as f_target:\n",
    "    #         pos = np.array(f_target['variants/POS'])\n",
    "    #         af = get_af1000G_atGivenPos(f_ref, pos)\n",
    "    # merge_in_af(path_h5, af)\n",
    "\n",
    "    # with h5py.File(f'./hdf5/ch{ch}.h5', 'a') as f:\n",
    "    #     # modify GP field to add a little bit uncertainty\n",
    "    #     nloci, nsample, _ = f['calldata/GT'].shape\n",
    "    #     gts = np.sum(f['calldata/GT'], axis=2)\n",
    "    #     assert(np.all(gts <=2))\n",
    "    #     gps = 0.0005*np.ones((nloci, nsample, 3))\n",
    "    #     for i in range(nsample):\n",
    "    #         for j in range(nloci):\n",
    "    #             gps[j, i, gts[j, i]] = 0.999\n",
    "    #     group = f['calldata']\n",
    "    #     group.create_dataset('GP', (nloci, nsample, 3), dtype='f')\n",
    "    #     f['calldata/GP'][:,:,:] = gps\n",
    "\n",
    "    print(f\"Finished running chromosome {ch}. GZ\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Run hapBLOCK on 4 samples...\n",
      "IBD1+IBD2 region total length: 109.376\n",
      "IBD2 region total length: 0.000\n",
      "IBD1+IBD2 region total length: 93.725\n",
      "IBD2 region total length: 0.000\n",
      "IBD1+IBD2 region total length: 101.158\n",
      "IBD2 region total length: 2.135\n",
      "IBD1+IBD2 region total length: 111.596\n",
      "IBD2 region total length: 2.044\n",
      "IBD1+IBD2 region total length: 86.793\n",
      "IBD2 region total length: 2.171\n",
      "IBD1+IBD2 region total length: 96.958\n",
      "IBD2 region total length: 0.000\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "sys.path.append(\"/mnt/archgen/users/yilei/tools/hapBLOCK/python3/\")     \n",
    "from run import hapBLOCK_all\n",
    "\n",
    "hapBLOCK_all(folder_in=\"./hdf5/ch\", iids = [], \n",
    "                   chs=range(1,23), folder_out=\"./hapBLOCK\", output=False, prefix_out=\"\", logfile=False,\n",
    "                   l_model=\"hdf5\", IBD2=True, p_col='default', \n",
    "                   ibd_in=1, ibd_out=10, ibd_jump=500, min_cm1=5, min_cm2=2,\n",
    "                   cutoff_post=0.99, max_gap=0.0075, save=3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# visualize "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append(\"/mnt/archgen/users/yilei/tools/hapBLOCK/python3/\")  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "findfont: Font family ['sans-serif'] not found. Falling back to DejaVu Sans.\n",
      "findfont: Font family ['sans-serif'] not found. Falling back to DejaVu Sans.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "plot IBD new version\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 432x288 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from plot.plot_posterior import plot_posterior_7States_plusGeno\n",
    "with open('merged.glimpse.GP99.MAC1.transversion_only.seg') as f:\n",
    "    counter = 1\n",
    "    for line in f:\n",
    "        id1, id2, ch, _, _, _, start, end, length, *_ = line.strip().split()\n",
    "        ch, start, end, length = int(ch), float(start), float(end), float(length)\n",
    "        plot_posterior_7States_plusGeno(f\"./hapBLOCK/{id1}_{id2}/{id1}_{id2}/chr{ch}\", start=start - 5, end=end + 5, \\\n",
    "            iids=[id1, id2], path2hdf5=\"./hdf5/ch\", ch=ch, prefix=str(counter), outFolder=\"./visualize\", truth=[start, end])\n",
    "        counter += 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to 0.99 GP variants: 1.000x\n",
      "Filtering to common GP variants: 1.000x\n",
      "[False False False False False False False False False  True  True False\n",
      " False False False False  True False False False False False False False\n",
      " False False False False  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 432x288 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "id1=\"I3950\"\n",
    "id2=\"I5273\"\n",
    "ch=10\n",
    "counter=71\n",
    "plot_posterior_7States_plusGeno(f\"./hapBLOCK/{id1}_{id2}/{id1}_{id2}/chr{ch}\", start=170, end=190, \\\n",
    "            iids=[id1, id2], path2hdf5=\"./hdf5/ch\", ch=ch, prefix=str(counter), outFolder=\"./visualize\")"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
  },
  "kernelspec": {
   "display_name": "Python 3.8.10 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
